[DOC] Add docs how to enable monitoring and alerting of the operator (#2764)
grafana/tempo: docs/sources/tempo/setup/operator/monitor.md

---
title: Monitor Tempo instances and the operator
description: Set up monitoring for Tempo instances and the operator
menuTitle: Monitor
weight: 300
aliases:
- /docs/tempo/operator/monitor
---

> # Monitor Tempo instances and the operator

Reviewer: Good title update! Thank you for updating the title in front matter as well.

Reviewer: Please feel free to edit that suggestion.

---
title: Monitor Tempo instances and the operator
description: Set up monitoring for Tempo instances and the operator
menuTitle: Monitor
weight: 300
aliases:
- /docs/tempo/operator/monitor
---

# Monitor Tempo instances and the operator

> You can configure the Tempo Operator to monitor TempoStack instances (including all Tempo components like the distributor). In addition, the operator can expose metrics about the operator itself (for example the number of successful and failed upgrades, etc.).

Reviewer: Adding a missing comma.
```suggestion
You can configure the Tempo Operator to monitor TempoStack instances (including all Tempo components like the distributor). In addition, the operator can expose metrics about the operator itself (for example, the number of successful and failed upgrades, etc.).
```

      sampling_fraction: "1.0"
      jaeger_agent_endpoint: localhost:6831
```

## Monitor the operator

> The Tempo Operator can expose upgrade and other operational metrics about the operator itself, and can create alerts based on these metrics.

Reviewer: What do you mean by "can expose upgrade and other operational metrics"?

Author: The operator handles Tempo upgrades, and it exposes metrics like "number of successful Tempo upgrades", "number of failed Tempo upgrades". And the operator create alerts, to notify system administrators if for example any Tempo upgrade failed.

The other operational metrics are internals of the operator, for example the duration of a "reconcile loop iteration". That's a operator-specific component which continuously tries to match the expected state as described in the TempoStack custom resource to the actual cluster state. E.g. if you delete some object in the cluster which is managed by the operator, the operator re-creates this object again, to match the expected state of the cluster.

Reviewer: This is great info!

> The configuration options to enable the creation of `ServiceMonitor` (for scraping metrics) and `PrometheusRule` (for creating alerts) objects is in the ConfigMap `tempo-operator-manager-config` in the same namespace as the operator:

Reviewer: This second sentence is a little confusing. Do you mean that someone has to add the example config map that you have below or that they need to set createServiceMonitors and createPrometheusRules to true in their ConfigMap?

Reviewer: Is this correct? 
```suggestion
To enable metrics, you need to create the `ServiceMonitor` (for scraping metrics) and `PrometheusRule` (for creating alerts) objects. These objects are configured in the ConfigMap `tempo-operator-manager-config` in the same namespace as the operator. 

To configure these objects, set `createServiceMonitors` and `crreatePrometheusRules` to `true`. 
```

Author: This config map already exists, it's part of the installation of the operator. This configmap contains various settings for the operator.

When the `createServiceMonitors` setting is enabled, the operator creates this `ServiceMonitor` object in the cluster, and another operator in the cluster (for example Prometheus Operator) reads this object and configures monitoring. The user does not create this `ServiceMonitor` object, this part is handled by the operator, if the `createServiceMonitors` setting is enabled.

I've rephrased this in my last commit, I hope this clears things up? If not, please let me know :)

Reviewer: Looks good! Thank you.

      sampling_fraction: "1.0"
      jaeger_agent_endpoint: localhost:6831
```

## Monitor the operator

The Tempo Operator can expose upgrade and other operational metrics about the operator itself, and can create alerts based on these metrics.

> The operator can be configured via the ConfigMap `tempo-operator-manager-config` in the same namespace as the operator.

Reviewer: ```suggestion
The operator can be configured using the ConfigMap `tempo-operator-manager-config` in the same namespace as the operator.
```
